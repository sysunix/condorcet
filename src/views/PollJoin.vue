<template>
  <Loader :processing="isProcessing">
    <div class="flex flex-col justify-center items-center h-full">
      <div class="text-2xl text-center w-full">{{ $t("poll.joined") }}</div>
      <div class="text-xl text-center w-full">{{ message }}</div>
    </div>
  </Loader>
</template>

<script lang="ts">
import { Vue, Component } from "vue-property-decorator";
import { State, Getter, Action, Mutation, namespace } from "vuex-class";
import { mapGetters } from "vuex";
import { db } from "../firebase";
import Loader from "../components/Loader.vue";

@Component({
  components: {
    Loader
  }
})
export default class PollJoin extends Vue {
  message: string = "";

  @Getter("user/userId") userId: any;

  get isProcessing() {
    return this.message === "";
  }

  async mounted() {
    try {
      const { params, query }: { params: any; query: any } = this.$route;
      const document = db.collection("polls").doc(params.id);
      const poll: any = (await document.get()).data();
      const { users, token: pollToken } = poll;

      if (!query.token) {
        return;
      }

      if (query.token !== pollToken) {
        return;
      }

      const temp: any = new Set([...users, this.userId]);
      const newPoll = { ...poll, users: [...temp] };

      document.set(newPoll);

      this.message = newPoll.question;
      setTimeout(() => {
        if (newPoll.isActive) {
          this.$router.push({ name: "poll_show", params: params.id });
          return;
        }

        this.$router.push({ name: "poll_result", params: params.id });
      }, 2000);
    } catch (error) {
      return;
    }
  }
}
</script>

<style></style>
