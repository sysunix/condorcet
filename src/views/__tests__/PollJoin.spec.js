import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import PollJoin from "../PollJoin.vue";

import {
  MOCK_POLL_ID,
  MOCK_TOKEN,
  MOCK_USER_ID,
  getTranslation
} from "../../utils/test";

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock("../../firebase");

const $route = {
  params: { id: MOCK_POLL_ID },
  query: { token: MOCK_TOKEN }
};

describe("PollJoin.vue", () => {
  let wrapper;

  beforeEach(() => {
    const store = new Vuex.Store({
      modules: {
        user: {
          namespaced: true,
          getters: {
            userId: () => MOCK_USER_ID
          }
        }
      }
    });

    wrapper = shallowMount(PollJoin, {
      localVue,
      store,
      mocks: { $route, $t: getTranslation }
    });
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
