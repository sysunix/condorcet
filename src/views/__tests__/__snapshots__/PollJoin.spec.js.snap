// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollJoin.vue should render 1`] = `
<loader-stub processing="true">
  <div class="flex flex-col justify-center items-center h-full">
    <div class="text-2xl text-center w-full">Tu as rejoint le scrutin :</div>
    <div class="text-xl text-center w-full"></div>
  </div>
</loader-stub>
`;
